### ISSUE TITLE
Cython compilation issues

### ISSUE THREAD

**Issue by mariavd** [qmcurrents/gimic#35]
I am having a nasty problem with Cython on some computers, including a new desktop I recently got. This computer runs CentOS 7. Anaconda 2.7 is already installed, and I also tried separately downloading Cython from their webpage, as well as using pip install:

```
[root@**** build]# which python
/home/****/anaconda2/bin/python

[root@**** build]# pip install cython
Requirement already satisfied: cython in /home/****/anaconda2/lib/python2.7/site-packages

[root@**** build]# which cython
/home/****/anaconda2/bin/cython
```

The end of the output from `make`:

```
[ 53%] Building CXX object src/libgimic/CMakeFiles/gimic2.dir/GimicInterface.cpp.o
[ 55%] Linking CXX shared library libgimic2.so
[ 55%] Built target gimic2
[ 57%] Compiling Cython CXX source for gimic...
Traceback (most recent call last):
  File "/home/****/anaconda2/bin/cython", line 11, in <module>
    load_entry_point('Cython==0.25.2', 'console_scripts', 'cython')()
  File "/home/****/anaconda2/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/pkg_resources/__init__.py", line 565, in load_entry_point
  File "/home/****/anaconda2/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/pkg_resources/__init__.py", line 2597, in load_entry_point
ImportError: Entry point ('console_scripts', 'cython') not found
make[2]: *** [src/libgimic/gimic.cxx] Error 1
make[1]: *** [src/libgimic/CMakeFiles/gimic.dir/all] Error 2
make: *** [all] Error 2
```

A similar problem also happens on our cluster in the lab and so far no solution has been found. Do you have any suggestions?

**Comment by bast** [comment]
Try temporarily deactivating Anaconda (by commenting respective call in your `.bashrc`), then try to install dependencies into a virtual environment:
```
virtualenv venv
source venv/bin/activate
pip install -r requirements.txt
```
Does that work?

My plan for the summer is to get rid of Cython in GIMIC.

**Comment by mariavd** [comment]
Thank you for the advice, I will try it tomorrow evening because I am finalizing a paper now. 

Otherwise, great that you will remove Cython. It will save lots of trouble. We need to have the compilation as smooth as possible.

**Comment by mariavd** [comment]
@bast So using the virtual environment let `cmake` create all the files in the build directory. It detected the system Python at /usr/bin:

```
-- Checking whether /usr/bin/gfortran supports Fortran 90 -- yes
-- Found Python: /usr/bin/python2.7  
-- Found Git: /usr/bin/git (found version "1.8.3.1") 
-- Found Cython: /usr/bin/cython  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/####/gimic/build
```

But then `make` ended up with an issue with Cython, too:

```
(venv) [root@#### build]# make
Scanning dependencies of target gimic2

...

Scanning dependencies of target gengauss
[ 65%] Building C object src/libgimic/CMakeFiles/gengauss.dir/gengauss.c.o
/home/####/gimic/build/src/libgimic/gengauss.c:436:31: fatal error: numpy/arrayobject.h: No such file or directory
 #include "numpy/arrayobject.h"
                               ^
compilation terminated.
make[2]: *** [src/libgimic/CMakeFiles/gengauss.dir/gengauss.c.o] Error 1
make[1]: *** [src/libgimic/CMakeFiles/gengauss.dir/all] Error 2
make: *** [all] Error 2

```

**Comment by bast** [comment]
I think I've had the same problem on Travis and this was my workaround: https://github.com/qmcurrents/gimic/blob/master/src/CMakeLists.txt#L1-L5 - it is a super hack but this problem will go away this summer when I replace Cython by CFFI or pybind11.

**Comment by mariavd** [comment]
An update. Adding the directory solves the numpy issue. Actually, even if the Anaconda binary is used it is possible to add this line to CMakeLists.txt to fix it.

`include_directories(/home/####/anaconda2/)`

Then Cython is causing an issue during `make`

```
[ 55%] Built target gimic2
[ 57%] Compiling Cython CXX source for gimic...
Traceback (most recent call last):
  File "/home/####/anaconda2/bin/cython", line 11, in <module>
    load_entry_point('Cython==0.25.2', 'console_scripts', 'cython')()
  File "/home/####/anaconda2/lib/python2.7/site-packages/pkg_resources/__init__.py", line 560, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/####/anaconda2/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2647, in load_entry_point
    raise ImportError("Entry point %r not found" % ((group, name),))
ImportError: Entry point ('console_scripts', 'cython') not found
make[2]: *** [src/libgimic/gimic.cxx] Error 1
make[1]: *** [src/libgimic/CMakeFiles/gimic.dir/all] Error 2
make: *** [all] Error 2
```

Including the directory did not help. Maybe it is a problem with the way Cython is set up? I asked about it in their mailing list. Even though it is installed and can be imported from the Python interpreter, if I just type `cython` in the command line the same error is printed:

```
[root@#### build]# conda install cython
Fetching package metadata .........
Solving package specifications: .   

# All requested packages already installed.
# packages in environment at /home/####/anaconda2:
#
cython                    0.25.2                   py27_0  
[root@#### build]# pip install cython
Requirement already satisfied: cython in /home/####/anaconda2/lib/python2.7/site-packages
[root@#### build]# which cython
/home/####/anaconda2/bin/cython
[root@#### build]# cython
Traceback (most recent call last):
  File "/home/####/anaconda2/bin/cython", line 11, in <module>
    load_entry_point('Cython==0.25.2', 'console_scripts', 'cython')()
  File "/home/####/anaconda2/lib/python2.7/site-packages/pkg_resources/__init__.py", line 560, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/####/anaconda2/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2647, in load_entry_point
    raise ImportError("Entry point %r not found" % ((group, name),))
ImportError: Entry point ('console_scripts', 'cython') not found
```

**Comment by mariavd** [comment]
@bast So, I somehow managed to get the compilation to work. No reply so far from the Cython developers about the above message. What I assume to have helped is that I did `yum install cython` and even though it installed Cython 0.19, it seems that something was fixed for the other cython installed by Anaconda. Then with the directory of numpy included in the CMakeLists.txt file, there were no longer problems. I am closing the comment and I believe there is no need to investigate the Cython issues any further since you will remove it. Thanks for the help.
